File: CPP/260.cpp
class Solution {
public:
	vector<int> singleNumber(vector<int>& nums) {
		int aXorB = 0;
		for (int each : nums) {
			aXorB ^= each;
		}
		/**
		 * lowbit is the minimal bit
		 * and in this bit, a is different from b.
		 */
		int lowbit = aXorB & -aXorB;
		int a = 0, b = 0;
		for (int each : nums) {
			if (each & lowbit) {
				a ^= each;
			} else {
				b ^= each;
			}
		}
		return vector<int> {a, b};
	}
};

File: Java/260.java
public class Solution {
    public int[] singleNumber(int[] nums) {
        int total = 0;
        for (int each : nums) {
            total ^= each;
        }
        int lowbit = total & -total;
        int a = 0, b = 0;
        for (int each : nums) {
            if ((each & lowbit) != 0) {
                a ^= each;
            } else {
                b ^= each;
            }
        }
        int[] ans = new int[2];
        ans[0] = a; ans[1] = b;
        return ans;
    }
}